### ISSUE TITLE
header-file not being output (silently).

### ISSUE THREAD

**Issue by MrBenGriffin** [Genivia/RE-flex#122]
Not really sure what is happening but neither 
--header-file=classic.h on the command line nor 
%option header-file=classic.h in the lpp file 
are generating anything, and I'm not being told why either.

I'm using the options

%option noyywrap unicode fast
%option bison-cc-namespace=yy
%option bison-cc-parser=CParser
%option bison-cc bison-locations noyywrap
%option namespace=yy
%option lexer="Classic"
%option header-file="classic.h"

I got everything working fine when I had just one parser/lexer - but outputting to 2 is not so easy - even location seems to have to be renamed.

I need the header so that the Bison parser(s) can undef yylex

#undef yylex
#define yylex classic.yylex  // bison's parse() should invoke lexer.yylex(), not yylex()

**Comment by MrBenGriffin** [comment]
This is only when running reflex under Xcode.  Maybe it's some environment header?

**Comment by genivia-inc** [comment]
This happens only under xcode? Perhaps the working directory is different? Is the scanner cpp source saved? Both are saved to the same current directory (unless header-file has a path). There is no environment variable that affects this.

**Comment by MrBenGriffin** [comment]
It’s a mystery that kept me awake all night…
I set the WD, copy all the files to it, run reflex … it generates the cpp … but nothing else.
If I run the script via shell (outside Xcode) it runs fine - and everything is generated.

I will try generate a basic package that replicates the issue…

Some sort of -v logging would be nice … I’m not expecting it though!

**Comment by genivia-inc** [comment]
It is never fun if something mysteriously fails. I am curious what could be the problem.

The `reflex` source code is pretty straightforward to save the header-file. Note: option header-file becomes the "header_file" key in the options[] map, because `-` is normalized to `_` to prevent confusion what to use and to allow `#define header_file` without syntax errors:
~~~
  if (!options["header_file"].empty())
  {
    ofs.open(options["header_file"].c_str(), std::ofstream::out);
    if (!ofs.is_open())
      abort("cannot open file ", options["header_file"].c_str());
    out = &ofs;
    *out <<
      "// " << options["header_file"] << " generated by reflex " REFLEX_VERSION " from " << infile << "\n\n" <<
      "#ifndef " << (prefix == "yy" ? "" : prefix.c_str()) << "REFLEX_" << upper_name(options["header_file"]) << '\n' <<
      "#define " << (prefix == "yy" ? "" : prefix.c_str()) << "REFLEX_" << upper_name(options["header_file"]) << '\n' <<
      "#define " << prefix << "IN_HEADER 1\n";
    write_prelude();
    ...
~~~

**Comment by genivia-inc** [comment]
Do you have instructions on how you set up your Xcode project so I can replicate this problem?

**Comment by MrBenGriffin** [comment]
Hi there Robert.
I enclose the package as a zip




I am using custom names because I need to include another Lexer, called ‘Classic’ 
They both generate identical tokens, and have identical syntaxes, except some of the reserved characters differ.

These are  ⌽⍟⎡⎤⎣⎦ - which, in classic, are @%{} ‘{|’ ‘|}’ respectively. 

I may be addressing this the wrong way - in the original, (written in flex), I declared an abstract lexer for the parser and the compiler/driver,
And maybe I should be doing the same here.

I did see that you suggest using different namespaces for different lexers - maybe I should try that.

Anyhow, here this is just the advanced lexer being used - and, for some reason, the headers are not appearing!

**Comment by genivia-inc** [comment]
Darn. I cannot open the Lexer.xcodeproj file because it is in a "future format". I'm at Xcode 12.4 (with updates installed.) I'm no fan of Big Sur. It tells me to "Adjust the project format using a compatible version of Xcode to allow it to be opened by this version of Xcode."

**Comment by MrBenGriffin** [comment]
This *says* it's Xcode 12 compatible...
[Uploading xc12.zip…]()

**Comment by genivia-inc** [comment]
Verified the build settings in the project editor and looks OK to me with my paths instead of yours:

<img width="611" alt="image" src="https://user-images.githubusercontent.com/13117347/153726880-f1751afc-5475-4eae-b92b-dc865456dd01.png">

However, this happens when building:

<img width="832" alt="image" src="https://user-images.githubusercontent.com/13117347/153726833-3dc25add-9ee0-406c-bd8c-f4fe751b6951.png">

My older Xcode is still trying to run Yacc, not bison.sh it seems.

On another note, may I suggest to change the reflex.sh script so that it exits with an error if reflex fails, which should get picked up by xcode:
~~~
/usr/local/bin/reflex --header-file=lex.${sbase}.h --outfile=${sbase}.yy.cpp ${sbase}.lpp || exit 1
~~~

**Comment by genivia-inc** [comment]
I can replicate the problem in xcode and see the same problem. Neither the advanced.yy.cpp nor the lex.advanced.h file are saved though.

When I remove `fast` from the `%option` in advanced.loop, then I got it to work once, I think, but not sure because I tried a lot of different scenarios for hours. For some reason, files cannot be written when the script runs or the reflex command fails to execute. The `fast` option appends the DFA code to the advanced.yy.cpp file after reflex generated the file initially. I saw that the TABLES are incomplete, suggesting that the file appending failed without warning. That is weird, but suggests that appending to a file fails also.

If I try the simplest script possible without command line arguments etc:
~~~
/usr/local/bin/reflex /Users/engelen/Desktop/xc12/Lexer/advanced.lpp 
~~~
I still do not get a lex.yy.cpp file. But when I run this on the command line, it works fine. I also tried with a very simple .l file, to see if it has something to do with the advanced.lpp file, but no success. Then I tried with a different newer (my dev) version of reflex that is faster, but still no success. Something in Xcode makes this script fail.

Or the Xcode configuration is not set correctly. In my case I don't see any evidence that the script runs, since I would expect before.log and after.log to show up, but it doesn't:
~~~
#/bin/sh
echo "OK" > /Users/engelen/Desktop/xc12/Lexer/before.log
/Users/engelen/Projects/reflex/src/reflex /Users/engelen/Desktop/xc12/Lexer/echo.l
echo "OK" > /Users/engelen/Desktop/xc12/Lexer/after.log
~~~

**Comment by MrBenGriffin** [comment]
Hi there,
I’m really so grateful that you have had a look at this.
I’m beginning to think it is my lack of understanding of XCode rather than reflex - but I am honestly mystified.

I notice that I must make clean (apple-shift-k) for the lexer to execute - which I believe to be that it’s generating advanced.yy.cpp

The 127 error code indicates “command not found”.  



I modified the script to 
--------
#!/bin/bash
src=${2##*/}
sbase=${src%%.*}
dir=${2%/*}
pdir=${3%/*}

cd ${pdir}
pushd ${dir}
cp ${3} .

echo "before: ${?} `date`" >> /tmp/reflex.log
/usr/local/bin/reflex -d --header-file=lex.${sbase}.h --outfile=${sbase}.yy.cpp ${sbase}.lpp 2&1 >> /tmp/reflex.log
echo "after: ${?} `date`" >> /tmp/reflex.log

popd
--------

Then using tail -f reflex.log which showed the shell above was working, but wouldn’t/couldn’t find reflex.

***@***.*** /tmp % tail -f reflex.log 
before: 0 Sun Feb 13 02:17:17 GMT 2022
after: 127 Sun Feb 13 02:17:17 GMT 2022


That 127 is interesting - apparently The 127 error code indicates “command not found”.
So I added library paths and paths - but it made no difference.

So I gave up (Again) on trying to use a shell script, and went back to using the custom setting

LEX=/usr/local/bin/reflex

This definitely does find it, and it runs - but there seems to be an issue (still) with generating the header file.
It shows duplicate options (but the errors seem to be appearing more reliably if the %option lines happen before %top )

Another issue with this is that it passes over (f)lex-like flags.

This is as far as I got today.

-B.

**Comment by MrBenGriffin** [comment]
Hi - my thinking was this…
The -o (from Xcode) includes the derived files directory, which is a really arbitrary path.  Adding that directory to the file therefore is very Xcode centric (I will be wanting to build in other systems).  I don’t want to bake the file name into the options in Xcode because it leaves just the one available, so I want the header-file option defined in the lexer file, but the header directory defined in Xcode settings…

**Comment by genivia-inc** [comment]
> Hi again, There’s a bit of an issue. I got some way by removing the handler script for reflex altogether. Defining LEX = /usr/local/bin/reflex Then using LEXFLAGS to set, eg --zenith and that gives a fail showing the values passed to reflex. It’s clear that Xcode depends upon lexer/parser files being generated at ${DERIVED_FILES_DIR} (This is available from the menu under “Project - show build folder”) So I’ve been monkey patching reflex.cpp to include a header-dir option (separating the header file from the directory) and running that alongside the values passed Something like

> ..but guess what? It’s still not showing!

There could be a difference with the %option fast. That option and %option full append the tables to the generated scanner. Just for now don't use these options and see what happens. If it works without these options, then there may be a problem in the table saving code located in the library source pattern.cpp. Perhaps the path is not set correctly. Right now, if pattern.cpp cannot append the file then it silently fails. We need an exception thrown though.

**Comment by MrBenGriffin** [comment]
Wow - well, I seem to have fixed it, but it may need some more investigation as to why!

I enclose the modified reflex.cpp - I moved the header_file output in front of the yy.cpp output (which meant I needed to move some of the initialisation stuff up too) - and it solved it.

My *total guess* is that there was something going wrong, but because the expected file was being generated, XCode took that as 'okay we are done here' - and just carried on.

I also include the (slightly modified) project code that compiles.

Ah.. I also modified the reflex lib makefile to get dylib working .. I added that here too.

[reflex.cpp.zip](https://github.com/Genivia/RE-flex/files/8056910/reflex.cpp.zip)
[Lexer.zip](https://github.com/Genivia/RE-flex/files/8056919/Lexer.z
[Make.zip](https://github.com/Genivia/RE-flex/files/8056929/Make.zip)
ip)

**Comment by MrBenGriffin** [comment]
a further thought on my prior guess.
I can imagine that Xcode is observing the output directory, waiting for the dependency it knows about (the -o file), and then considers itself able to continue processing the very moment that file is generated (and the file handle is closed).  For all I know, it may then kill the generating processor.  

If there's any basis to this thought, then it does make a lot of sense (to me) to generate all ancillary files and secondary files before the primary output file.

**Comment by genivia-inc** [comment]
> a further thought on my prior guess. I can imagine that Xcode is observing the output directory, waiting for the dependency it knows about (the -o file), and then considers itself able to continue processing the very moment that file is generated (and the file handle is closed). For all I know, it may then kill the generating processor.
> 
> If there's any basis to this thought, then it does make a lot of sense (to me) to generate all ancillary files and secondary files before the primary output file.

Perhaps that is the case. But generating the header file first will not help, because the .cpp file is saved and afterwards opened again to append the tables.

Is it possible to make the dependency in Xcode on the .h file since it is generated last, instead of the .cpp file?

**Comment by MrBenGriffin** [comment]
I wouldn’t know how to - as lex / flex generate cpp files and that is what Xcode is looking for.. 

I honestly know very little how you work the inner magic - so you may be right… I was just happy to see something generate and compile!

**Comment by genivia-inc** [comment]
Another experiment that is worth a try to figure this out: let reflex generate a different .cpp file in the script. Then rename the file after reflex is finished with `mv` in the same script.

**Comment by genivia-inc** [comment]
I don't know if this issue came to any conclusion how to move forward? I'm closing this until this is perhaps more clear what the Xcode internals are with respect to custom builds steps to generate source code.

**Comment by MrBenGriffin** [comment]
The choice is yours, of course - however the problem remains - that XCode uses a unique DERIVED_FILES_DIR; this is important because it enables the output headers to be placed in the correct place. - (which are otherwise placed in the project directory) as for the mystery regarding why outputting the headers before the cpp, my guess is that it has something to do with concurrency and dependency management - but that’s as far as I got.


In brief, though, there is no good example to follow for building reflex under Xcode - and that is truly a disappointment.

**Comment by genivia-inc** [comment]
It is not clear to me what the Xcode logic is. If it's based on the file write and close to kill the process afterwards, then you could just change the shell script to let reflex generate the files first, but with a different filename. Then copy the scanner source code file to the filename that Xcode expects. If that works, great! If that does not work, Xcode does something different than we think it is doing.
